[BEAM-2857] Implementing WriteToFiles transform for fileio (Python)

pabloem opened the pull request:
r: @chamikaramj

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PreCommit

pabloem commented:
Passing postcommit: https://builds.apache.org/job/beam_PostCommit_Python_Verify_PR/593/
PAssing py3 postcommit: https://builds.apache.org/job/beam_PostCommit_Python3_Verify_PR/230/

udim commented:
still reviewing...

pabloem commented:
ugh... I'm so sorry Udi. The design doc for this is here: https://s.apache.org/fileio-beam-python
it's light on details.. but i guess... it would be helpful : /

udim commented on sdks/python/apache_beam/transforms/core.py:
```diff
@@ -1982,7 +1983,11 @@ def process(self, element, timestamp=DoFn.TimestampParam,
+               timestamp_combiner=None):
     """Initializes a WindowInto transform.
 
     Args:
```
Could you add the allowable types for `windowfn` (Windowing and WindowFn?).

udim commented on sdks/python/apache_beam/testing/util.py:
```diff
@@ -141,6 +145,21 @@ def _equal(actual):
   return _equal
 
 
+def matches_all(expected):
```
Is this different from `contains_in_any_order` above?

pabloem replied:
The `contains_in_any_order` seems to require a set of hashable items. It simply does `collections.Counter()` on the iterable, and compares both counters. `matches_all`  takes in an iterable of elements or matchers, and matches them 1:1 properly.

udim replied:
gotcha

pabloem commented on sdks/python/apache_beam/io/fileio.py:
```diff
@@ -169,3 +179,476 @@ def __init__(self, compression=None, skip_directories=True):
+          str,
+          filesystems.FileSystems.join(temp_location,
+                                       '.temp%s' % dir_uid))
+      logging.info('Added temporary directory %s', self._temp_directory.get())
```
No. The transform may be run in streaming, so I don't think there's a good moment in the pipeline execution to perform a cleanup of the temp directory.

pabloem commented:
Hi Udi. I've addressed some of your comments. Some things pending to finalize:

- How do we clean up temporary files?
- In general, how do we deal with failures?

Some things that are worth looking at:

- I added documentation at the top of the file. LMK if that clarifies some things.
- Default sink is now a text sink that encodes strings to UTF8 encoding.

pabloem commented:
Run Python PreCommit

pabloem commented:
I've updated the cleanup code. Any error other than `BeamIOError` should be a retry. Also, I believe it should be okay to only clean up all of the files, and not the directory (as we'd have to check for batch vs streaming).
Thoughts?

udim commented:
From what I understand deletion of temporary files is best effort. There is currently no mechanism to clean up old temporary files left over due to an error. Perhaps we could, in the future, add a TTL or some other way to tell if a temporary file can be safely deleted.

udim commented on sdks/python/apache_beam/io/fileio.py:
```diff
@@ -554,6 +561,18 @@ def process(self, element):
 
+    logging.info('Cautiously removing temporary files for destination %s',
+                 destination)
+    self._remove_temporary_files([r.file_name for r in file_results])
```
There's an implicit `return None` at the end of this method. Are `None`s handled correctly?

pabloem replied:
I believe this method will return a generator due to the above yield keyword, so this does not return None.

udim commented:
Looks good, except for the new fileio.py comment I left above.

pabloem commented:
> From what I understand deletion of temporary files is best effort. There is currently no mechanism to clean up old temporary files left over due to an error. Perhaps we could, in the future, add a TTL or some other way to tell if a temporary file can be safely deleted.

Do you mean when the pipeline fails? (e.g. for batch?)

You're right about that, that if the pipeline fails, we won't be able to clean up the bundles that did not run. But how could we know this ahead of time? : /

Currently, all files will be cleaned up for any bundle that succeeds. If the bundle does not succeed, then we should not clean up the files.

pabloem commented:
Run Python_PVR_Flink PreCommit

udim commented:
Yes, it applies to cases where the pipeline fails, but could also be in cases where it is cancelled or updated (does the temp directory change in that case?).

We could format temp directory names to have the job ID in them, so it's possible to check if the job is still running and delete temp files if it isn't. This process would not be run as part of the pipeline.
Regarding temp directory deletion, I believe that for batch jobs they should be removed at the end (because it's better to clean up if we can). For streaming jobs, there's no hook for cleanup during cancellation and updates AFAIK.

Regarding failed bundles, assuming that the pipeline has not failed, it should be possible to scan for and delete any remaining temporary files.

udim commented:
cc: @chamikaramj as I'm OOO next week.

pabloem commented:
@chamikaramj

chamikaramj commented on sdks/python/apache_beam/io/fileio.py:
```diff
@@ -502,12 +502,13 @@ def expand(self, pcoll):
 
-def _create_writer(base_path, file_name):
+def _create_writer(base_path, writer_key):
   # TODO(pabloem) Is this a good place to create a temporary directory?
```
Please create a JIRA or remove this.

pabloem replied:
Done.

chamikaramj commented on sdks/python/apache_beam/io/fileio.py:
```diff
@@ -562,16 +569,21 @@ def process(self, element):
+      orphaned_files = [m.path for m in match_result[0].metadata_list]
+
+      logging.debug('Deleting orphaned files: %s', orphaned_files)
+      filesystems.FileSystems.delete(orphaned_files)
```
Shouldn't we be ignoring missing source files to handle workitem retries ?

udim replied:
I agree that we should ignore errors here. Note that `orphaned_files` are taken from a recent `filesystems.FileSystems.match()` call, so it doesn't make a difference if this is a retry.

pabloem replied:
Yes. All we're doing is logging the failures (in debug level) to be aware of them.

chamikaramj commented on sdks/python/apache_beam/io/fileio.py:
```diff
@@ -170,3 +237,487 @@ def __init__(self, compression=None, skip_directories=True):
+      kwargs['start'] = window.start.to_utc_datetime().isoformat()
+      kwargs['end'] = window.end.to_utc_datetime().isoformat()
+
+    # TODO(pabloem): Add support for PaneInfo
```
Create a JIRA and mention here ?

pabloem replied:
Done

chamikaramj commented on sdks/python/apache_beam/io/fileio.py:
```diff
@@ -23,17 +23,83 @@
+File Naming
+-----------
+One of the parameters received by ``WriteToFiles`` is a function specifying how
+to name the files that are written. This is a function that takes in the
```
Also document that this is supported for streaming and how it is supported.

pabloem replied:
Done.

chamikaramj commented:
LGTM. Thanks.

Please fix-up and self-merge after tests pass.

pabloem commented:
Run Python_PVR_Flink PreCommit

chamikaramj commented:
Also, please fix the conflict.

pabloem commented:
Run Portable_Python PreCommit

pabloem commented:
Run Python PreCommit

yupbank commented:
Hey, So following the doc for which `AvroSink`  clearly doesn't exist and the `beam.io.avroio._AvroSink` would require much more info. For the context, i'm looking into ways to do `WriteToParquet` from a pubsub input. and `WriteToFiles` provides the closest thing. But the document is a bit confusing.
